All Sections to SubSections

hahmadia opened the pull request:
#### Summary
<!--
A description of what this pull request does.
-->
Ultimate PR to merge all Sections-To-SubSection PR's.
The goal of this PR is to merge all of the section-to-subsection migrations at once into master.
We do not want to merge each PR one at time into master (due to cloud release and it hence it will seem incomplete), we've created this branch which all the individual PR's can be merged into and then this PR can be merged in at once into master.

This branch includes all the following:

**Experimental:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7475
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/16887
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1356

**About:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7593
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17015
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1375

**Reporting:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7615
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17035
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1386


**Environment:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7622
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17054
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1389


**Site Configuration:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7666
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17095
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1403


**Authentication:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7660
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17087
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1401

**Integrations:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7671
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17097
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1404

**Compliance:**

- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7596
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17023
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1377

### Other PR's required

**Determining Ancillary Permissions on Backend rather than Frontend:**
- [x] Webapp: https://github.com/mattermost/mattermost-webapp/pull/7628
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17061
- [x] Redux: https://github.com/mattermost/mattermost-redux/pull/1392

### PR's required to be merged before this
- [x]  Server: https://github.com/mattermost/mattermost-server/pull/17163

hahmadia commented:
/update-branch

hahmadia commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

hahmadia commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

hahmadia commented:
/update-branch

hahmadia commented:
/update-branch

hahmadia commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

furqanmlk commented:
1. **On Read-only permission I can Choose the file to upload in ALDAP section but upload button does nothing.
   In SAML section these buttons are disabled**

![image](https://user-images.githubusercontent.com/6909890/112211493-56d3ce00-8bf2-11eb-815d-b3fa81dbf235.png)

**2. When **team** / **channel** permission is ```Read only``` or ```No Access``` for System Manager, UI is disable but user can be added and Removed from **team/channel** through API while  **

![image](https://user-images.githubusercontent.com/6909890/112228062-09159080-8c07-11eb-8e3c-6c6be6a389c8.png)

furqanmlk reviewed (commented):
Testing still in progress . . .

BenCookie95 commented:
One thing I noticed is for pages that aren't built using `admin_definition` (E.G Data Retention has it's own component) I can still click buttons if I have read access (Try click the run job button on data retention). Obviously the server rejects the request but there is no UI feedback of the request failing and it's just inconsistent to other pages where the button would be disabled. 

I wonder if we need to do a permission checks on all the buttons on those custom system console pages in order to give a consistent UX?

BenCookie95 commented:
Sections that have custom components and buttons that needs permissions:
```
Data Retention
Compliance Export
Bleve
```

mkraft commented on components/admin_console/admin_definition.jsx:
```diff
@@ -401,11 +420,11 @@ const AdminDefinition = {
+            it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.USER_MANAGEMENT.GROUPS)),
+            it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.USER_MANAGEMENT.TEAMS)),
+            it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.USER_MANAGEMENT.CHANNELS)),
+            it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.USER_MANAGEMENT.PERMISSIONS)),
```
It would be great if these arrays of sub-permissions updated these parent sections from a single source data structure to make it impossible to forget to add one. 

Actually, maybe we should switch to a `permissions` parameter that handled the imperative part of the logic (namely, disabling and or hiding). That way a developer making a change to these menu items later doesn't have to think about any of `not`, `all`, `isDisabled`, `isHidden`, and `userHasReadPermissionOnResource` things.

mkraft replied:
@hahmadia as we discussed offline, I created a ticket for us to handle this refactor: https://mattermost.atlassian.net/plugins/servlet/mobile?originPath=%2Fbrowse%2FMM-34281#issue/MM-34281

furqanmlk commented:
User with “No Access” permission on “User Management” can still access via url User/Teams/Groups/Channel page and view
```
https://localhost:8065/admin_console/user_management/groups/<group-id>
https://localhost:8065/admin_console/user_management/teams/<team-id>
https://localhost:8065/admin_console/user_management/channels/<channel-id>
https://localhost:8065/admin_console/user_management/users/<user-id>
```
![image](https://user-images.githubusercontent.com/6909890/112388827-20698200-8cca-11eb-88d6-c80ee289861b.png)

hahmadia commented on components/admin_console/admin_definition.jsx:
```diff
@@ -224,7 +230,9 @@ const AdminDefinition = {
         isHidden: it.any(
             it.configIsTrue('ExperimentalSettings', 'RestrictSystemAdmin'),
-            it.not(it.userHasReadPermissionOnResource('about')),
+            it.all(
```
Oops.

hahmadia commented on components/admin_console/admin_definition.jsx:
```diff
@@ -240,7 +248,8 @@ const AdminDefinition = {
             ],
-            isDisabled: it.not(it.userHasWritePermissionOnResource('about')),
+            isHidden: it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.ABOUT.EDITION_AND_LICENSE)),
+            isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ABOUT.EDITION_AND_LICENSE)),
```
The section gets hidden when we doing have read permission and disabled when we don't have write permission. The disabling is for read only admins. The hiding is for no access admins. There is buttons on that page like upload license that need disable when your a read only admin.

hahmadia commented:
@furqanmlk 
I've addressed the bugs not related to user management section. As discussed offline, tickets will be created for the bugs found in the user management section to be addressed at a later time.

@BenCookie95 
We spoke offline and neither of us could reproduce the bugs mentioned. Scott and I did discover a small button not being disabled on compliance monitoring section which I have addressed.

@mkraft 
We spoke offline and we came to the conclusion that the bigger refactor Martin had in mind should be done at a later date. He created a ticket for it (https://mattermost.atlassian.net/browse/MM-34281). As for the section `isHidden` functions, I've refactored down to a `isHidden: it.not(it.userHasReadPermissionOnResources(Resource_Keys.Section_name))` which goes through all of it's permissions checking to see if we have read permission on each. This way, we eliminate having to list out each permission individually.

BenCookie95 reviewed (approved):
LGTM now, nice job!

furqanmlk reviewed (approved):
Issued #01 has been tested, working as expected.
I have created different tickets for other two issues.
Everything else looks good

sbishel commented:
/update-branch

ethervoid commented:
/cherry-pick cloud

mattermod commented:
Cherry pick is scheduled.
